pkg/kubectl/cmd/expose.go

```diff
@@ -149,101 +137,124 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 		return cmdutil.UsageError(cmd, fmt.Sprintf("generator %q not found.", generatorName))
 	}
 	names := generator.ParamNames()
-	params := kubectl.MakeParams(cmd, names)
-	name := info.Name
-	if len(name) > validation.DNS952LabelMaxLength {
-		name = name[:validation.DNS952LabelMaxLength]
+
+	// Get the input object
+	inputObject, err := r.Object()
+	if err != nil {
+		return err
 	}
-	params["default-name"] = name
 
-	// For objects that need a pod selector, derive it from the exposed object in case a user
-	// didn't explicitly specify one via --selector
-	if s, found := params["selector"]; found && kubectl.IsZero(s) {
-		s, err := f.MapBasedSelectorForObject(inputObject)
-		if err != nil {
-			return cmdutil.UsageError(cmd, fmt.Sprintf("couldn't retrieve selectors via --selector flag or introspection: %s", err))
-		}
-		params["selector"] = s
+	var inputObjects []runtime.Object
+	switch inputObject.(type) {
+	case *api.List:
+		inputObjects = inputObject.(*api.List).Items
+	default:
+		inputObjects = append(inputObjects, inputObject)
 	}
 
-	// For objects that need a port, derive it from the exposed object in case a user
-	// didn't explicitly specify one via --port
-	if port, found := params["port"]; found && kubectl.IsZero(port) {
-		ports, err := f.PortsForObject(inputObject)
-		if err != nil {
-			return cmdutil.UsageError(cmd, fmt.Sprintf("couldn't find port via --port flag or introspection: %s", err))
+	for i := range infos {
+		info := infos[i]
+		mapping := info.ResourceMapping()
+		if err := f.CanBeExposed(mapping.GroupVersionKind.GroupKind()); err != nil {
+			return err
 		}
-		switch len(ports) {
-		case 0:
-			return cmdutil.UsageError(cmd, "couldn't find port via --port flag or introspection")
-		case 1:
-			params["port"] = ports[0]
-		default:
-			params["ports"] = strings.Join(ports, ",")
+
+		params := kubectl.MakeParams(cmd, names)
+		name := info.Name
+		if len(name) > validation.DNS952LabelMaxLength {
+			name = name[:validation.DNS952LabelMaxLength]
 		}
-	}
-	if kubectl.IsZero(params["labels"]) {
-		labels, err := f.LabelsForObject(inputObject)
+		params["default-name"] = name
+
+		// For objects that need a pod selector, derive it from the exposed object in case a user
+		// didn't explicitly specify one via --selector
+		if s, found := params["selector"]; found && kubectl.IsZero(s) {
+			s, err := f.MapBasedSelectorForObject(inputObjects[i])
```

Comment:
sometimes using `infos` and sometimes using `inputObjects` seems strange.  What's going on that requires this?